Basic Mob Brimdemon

Jacquerel opened the pull request:
## About The Pull Request

Fixes #71330

The brimdemon was basically already perfect (well, it has a novel means of attacking) so I didn't get too fancy with this one, it's _largely_ just a straightforward conversion.
Following this change it's a little slower to back off, but better at lining up with people in order to blast them. Additionally, its beam is now a mob ability so you can give it to other mobs if you so desire.

Because I can't help doing a _little_ tinkering, Brimdemons now explode 2.5 seconds after they die, after a brief warning animation.

## Why It's Good For The Game

Simple mobs must die

## Changelog

:cl:
add: Brimdemon corpses release an explosion shortly after death, just to keep you on your toes.
refactor: Brimdemons now use the basic mob framework which (should) improve their pathfinding somewhat. Please bug report any unusual behaviour.
admin: The brimdemon's beam ability can be given to any mob, for your Binding of Isaac event
/:cl:

Jacquerel committed:
Match player and AI melee attack cooldowns

Jacquerel committed:
Faithless behaviour > attack

Jacquerel committed:
Elementise behaviour, share with lobstrosity

Jacquerel committed:
star gazer & ice whelp behaviour > attack
blocked isnt a boolean

Jacquerel committed:
use grabbedby instead & fix inverted check
remove mob_grabber from star gazer because its completely pointless (it leaves a trail which blocks the mobs it has grabbed?)

Jacquerel committed:
Base stats and Brimbeam

Jacquerel committed:
Just use nextmove instead of a weird extra system

Jacquerel committed:
this comment isnt true

Jacquerel committed:
Merge branch 'fist-of-the-carp' into to-cinders

Jacquerel committed:
Death burst

Jacquerel committed:
AI

Jacquerel committed:
Tuning & old file removal

Jacquerel committed:
Merge branch 'master' into to-cinders

Jacquerel committed:
that was close i almost let san tell me to alphabetise some lists

Jacquerel committed:
update update paths paths

SyncIt21 commented:
Fixes #71330

Jacquerel committed:
naughty renamed kwargs

san7890 reviewed (changes requested):
nits and an optional thing

Jacquerel commented on code/datums/ai/idle_behaviors/idle_random_walk.dm:
```diff
@@ -14,3 +14,13 @@
+	var/target_key = BB_BASIC_MOB_CURRENT_TARGET
+
+/datum/idle_behavior/idle_random_walk/no_target/perform_idle_behavior(seconds_per_tick, datum/ai_controller/controller)
+	if (!QDELETED(controller.blackboard[target_key]))
```
We do this often enough that it seems sensible

Jacquerel committed:
San has unleashed a significant increase in my touched file count

Jacquerel committed:
move this down one line

san7890 commented on code/modules/mob/living/basic/lavaland/brimdemon/brimdemon.dm:
```diff
@@ -65,8 +65,8 @@
 /obj/effect/temp_visual/brim_burst/Initialize(mapload)
-	addtimer(CALLBACK(src, PROC_REF(bang)), duration - 1, TIMER_DELETE_ME)
 	. = ..()
+	addtimer(CALLBACK(src, PROC_REF(bang)), duration - 1, TIMER_DELETE_ME)
```
where's my deciseconds define

Ben10Omintrix commented on code/datums/ai/basic_mobs/basic_ai_behaviors/targeted_mob_ability.dm:
```diff
@@ -23,7 +25,7 @@
 		return
 	var/mob/living/living_target = target
-	if(living_target.stat >= UNCONSCIOUS)
+	if(living_target.stat >= minimum_stat)
```
maybe this could read the stat_attack from the targetting datum instead?

Jacquerel committed:
Sure let's multiply this number by one
Refactor behaviour into a subtype of existing behaviour
Add "don't do anything during an action bar" as an ai flag
Removes targetted ability code which actually didn't do anything because this is already handled by the find target action

Jacquerel committed:
loadbearing lazylen

Shadowflame909 commented:
So close for the allmighty var edit tool allowing admins to give me my 30 damage kamehameha (its just a var-editted brimdemon beam)

Ben10Omintrix commented on code/modules/mob/living/basic/lavaland/goldgrub/goldgrub_ai.dm:
```diff
@@ -108,10 +108,15 @@
 
 	//a storm is coming or someone is nearby, its time to escape
-	if(currently_underground || !currently_underground && storm_approaching || !QDELETED(target))
+	if(currently_underground)
```
1 small thingy this would mean that during storms, goldgrubs would keep digging out and in. my intention was the grub will remain underground until the storm passes but if u want to change it then its fine

Jacquerel replied:
Worry not it calls this a few lines above:
```
if(currently_underground && storm_approaching)
		return SUBTREE_RETURN_FINISH_PLANNING
```

It _does_ act pretty weird when burrowing to escape someone (on live right now) though, basically doing the popping up and down thing you mention here. I think it loses its target after burrowing or something, was going to make a future PR to investigate 😔

Ben10Omintrix replied:
my mistake i didnt see that
